lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api.php

```diff
@@ -0,0 +1,668 @@
+		 * The suffix is the string after the first double hyphen and the prefix is
+		 * everything that comes before the suffix.
+		 *
+		 * Example:
+		 *
+		 *     extract_prefix_and_suffix( 'data-wp-interactive' )   => array( 'data-wp-interactive', null )
+		 *     extract_prefix_and_suffix( 'data-wp-bind--src' )     => array( 'data-wp-bind', 'src' )
+		 *     extract_prefix_and_suffix( 'data-wp-foo--and--bar' ) => array( 'data-wp-foo', 'and--bar' )
+		 *
+		 * @since 6.5.0
+		 *
+		 * @param string $directive_name The directive attribute name.
+		 * @return array An array containing the directive prefix and optional suffix.
+		 */
+		private function extract_prefix_and_suffix( string $directive_name ): array {
+			return explode( '--', $directive_name, 2 );
+		}
+
+		/**
+		 * Parses and extracts the namespace and reference path from the given
+		 * directive attribute value.
+		 *
+		 * If the value doesn't contain an explicit namespace, it returns the default
+		 * one. If the value contains a JSON object instead of a reference path, the
+		 * function tries to parse it and return the resulting array.
+		 *
+		 * Example:
+		 *
+		 *     extract_directive_value( 'actions.foo', 'myPlugin' )                      => array( 'myPlugin', 'actions.foo' )
+		 *     extract_directive_value( 'otherPlugin::actions.foo', 'myPlugin' )         => array( 'otherPlugin', 'actions.foo' )
+		 *     extract_directive_value( '{ "isOpen": false }', 'myPlugin' )              => array( 'myPlugin', array( 'isOpen' => false ) )
+		 *     extract_directive_value( 'otherPlugin::{ "isOpen": false }', 'myPlugin' ) => array( 'otherPlugin', array( 'isOpen' => false ) )
+		 *
+		 * @since 6.5.0
+		 *
+		 * @param string|true $directive_value   The directive attribute value. It can be `true` when it's a boolean
+		 *                                       attribute.
+		 * @param string|null $default_namespace Optional. The default namespace if none is explicitly defined.
+		 * @return array An array containing the namespace in the first item and the JSON, the reference path, or null on the
+		 *               second item.
+		 */
+		private function extract_directive_value( $directive_value, $default_namespace = null ): array {
+			if ( empty( $directive_value ) || is_bool( $directive_value ) ) {
+				return array( $default_namespace, null );
+			}
+
+			// Replaces the value and namespace if there is a namespace in the value.
+			if ( 1 === preg_match( '/^([\w\-_\/]+)::./', $directive_value ) ) {
+				list($default_namespace, $directive_value) = explode( '::', $directive_value, 2 );
+			}
+
+			/*
+			 * Tries to decode the value as a JSON object. If it fails and the value
+			 * isn't `null`, it returns the value as it is. Otherwise, it returns the
+			 * decoded JSON or null for the string `null`.
+			 */
+			$decoded_json = json_decode( $directive_value, true );
+			if ( null !== $decoded_json || 'null' === $directive_value ) {
+				$directive_value = $decoded_json;
+			}
```

Comment:
it'd still be nice to clarify what the expectations are here for sending non-object JSON. what happens if we store `"something"` as the value, or `false` or `true`, or `[{"o":4}, null, 3e6]`

Reply:
Sure! Done in https://github.com/WordPress/gutenberg/pull/58066/commits/6b62b970c7986e457e3bb3d848292a9072653dfe and added more edge cases in https://github.com/WordPress/gutenberg/pull/58066/commits/93eefeb37184cca608c23c641661f7031f0f9c45.

Reply:
thanks @luisherranz - I think I poorly worded what I was trying to convey. I'm putting myself into the shoes of someone who's wanting to understand the role that JSON data plays in the directives. maybe this is all documented somewhere else (in which case a link would be great).

in all of the iterations until now, the function's docblock has given examples of some JSON inputs and what it will do with those, but even as we add new examples, it still leaves open the question of the other kinds of JSON input not listed.

it would be great to somewhat comprehensively indicate what role the JSON plays. right now, it seems like we allow all JSON values except for `null` and `"null"`. I will therefore assume that I can store a JSON array like the value of `d` in `d = [null, null, null]` as the directive value. there's no indication in the documentation what this should produce. it's valid JSON but it's neither a JSON object, nor a number, nor the string `"null"` nor `"false"` nor `"true"`.

back to the comment, I see two ways of changing this that would clarify the expectations, both involve swapping out _some_ examples of JSON inputs and replacing them with the specification for it.

1. match the docblock to the existing code

```php
/*
 * The value may be one of the following:
 *
 *  - a namespace-prefixed path into the store, e.g. `otherPlugin::actions.foo`.
 *  - a dot-separated path without a namespace, e.g. `actions.foo`.
 *  - any JSON-serialized value except `null` and the string `"null"`, e.g. `{"isOpen": false}`.
 *
 * If provided an invalid value, or a value that cannot be decoded as JSON
 * the original value will pass through and fail to refer to any values from
 * the store.
 *
 * Non-string JSON values will be coerced to strings in the markup.
 */
```

2. match the code to the existing docblock

```php
$decoded_json = json_decode( $directive_value, /* Disambiguate arrays from objects. */ false );
if (
	"true" === $decoded_json ||
	"false" === $decoded_json ||
	is_object( $decoded_json ) ||
	is_numeric( $decoded_json )
) {
	$directive_value = is_object( $decoded_json )
		? (array) $decoded_json
		: (string) $decoded_json;
}
```

I _think_ that the second approach more closely matches what you were intending based on the function's explanation, examples, and tests. I may be totally confused in part because I don't understand what it means to "parse it and return the resulting array." What does it mean to pass `{"isOpen": false}`? or `otherPlugin::{"isOpen": false}`? would this be valid? `otherPlugin::[null]`?

please receive this comment in a tone of curiosity and good-will. I'm not trying to demand you change this, but it remains one of the more confusing points for me to understand and I'm sharing because I would imagine that others could potentially have similar questions. I also understand that this is an internal method and not the public interface, but I would still find it valuable to know what's supposed to happen if I am pulled in to debug an issue with this code and have a JSON input that isn't specified in the function's docs.